### PR TITLE
[ONNX] Remove usage of isCompleteTensor() in symbolic functions

### PR DESCRIPTION
**Pull request by BowenBao**
`isCompleteTensor()` only returns true when both scalar type and shape is present. All dimensions in the shape must be static. This high requirement is unnecessary for many use cases such as when only rank or scalar type needs to be known.

### CODE REVIEW

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -177,6 +177,17 @@ def _is_tensor(x):
...
+    if not _is_tensor(x) or x.type() is None:
+        return None
+    return x.type().dim()
+
+def _get_tensor_sizes(x, allow_symbol=True):
+    if not _is_tensor(x) or x.type() is None:
+        return None
+    if allow_symbol:
+        return x.type().varyingSizes()
+    return x.type().sizes()
```
Comment (jiafatom) [inline torch/onnx/symbolic_helper.py:190]: If `allow_symbol`, the size contains symbols? If not, then the dynamic size would return `None` for that dimension?

File: torch/onnx/symbolic_opset8.py [hunk 2]
```diff
@@ -148,10 +148,9 @@ def matmul(g, self, other):
 
 
 def prelu(g, self, weight):
-    if self.isCompleteTensor():
-        self_sizes = self.type().sizes()
-        if self_sizes and len(self_sizes) > 2:
-            weight = g.op("Unsqueeze", weight, axes_i=list(range(1, len(self_sizes) - 1)))
+    self_rank = sym_help._get_tensor_rank(self)
+    if self_rank and self_rank > 2:
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset8.py:152]: `self_rank is not None` for consistency?

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -474,8 +474,9 @@ def transpose(g, self, dim0, dim1):
         return self
 
     # NB: Transpose in ONNX is actually a Permute
-    if self.isCompleteTensor():
-        axes = list(range(self.type().dim()))
+    rank = sym_help._get_tensor_rank(self)
+    if rank:
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset9.py:478]: ditto

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -638,10 +639,9 @@ def squeeze(g, self, dim=None):
     return g.op("Squeeze", self, axes_i=[squeeze_dim])
 
 def prelu(g, self, weight):
-    if self.isCompleteTensor():
-        self_sizes = self.type().sizes()
-        if self_sizes and len(self_sizes) > 2:
-            weight = g.op("Unsqueeze", weight, axes_i=list(range(1, len(self_sizes) - 1)))
+    self_rank = sym_help._get_tensor_rank(self)
+    if self_rank and self_rank > 2:
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset9.py:643]: ditto

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -881,11 +880,12 @@ def symbolic_fn(g, input, output_size):
             return sym_help._onnx_unsupported('adaptive pooling, since output_size is not constant.')
         if output_size == [1] * len(output_size) and type == "AveragePool":
             return g.op("GlobalAveragePool", input)
-        if not input.isCompleteTensor():
+        sizes = sym_help._get_tensor_sizes(input)
+        dim = sizes[2:] if sizes is not None else None
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset9.py:884]: need `len(sizes) > 2`?

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -1298,13 +1298,14 @@ def instance_norm(g, input, weight, bias, running_mean, running_var, use_input_s
 def unfold(g, input, dimension, size, step):
     if sym_help._operator_export_type == torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK:
         return g.op("ATen", input, operator_s="unfold", dimension_i=dimension, size_i=size, step_i=step)
-    if input.isCompleteTensor():
-        sizedim = input.type().sizes()[dimension]
+    sizes = sym_help._get_tensor_sizes(input)
+    sizedim = sizes[dimension] if sizes is not None else None
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset9.py:1302]: check index out of boundary?

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:194]: added comments for details.

**Review by jiafatom — APPROVED** [review]
LGTM!

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -523,16 +526,21 @@ def prim_ConstantSplit(g, self, split_size, dim):
...
     return prim_ConstantSplit(g, self, split_size, dim)
 
 
 @parse_args('v', 'i', 'i', 'i')
 def unsafe_chunk(g, self, chunks, dim, _outputs=None):
     if _outputs is None:
         return sym_help._onnx_opset_unsupported_detailed('unsafe_chunk', 9, 11, 'Dynamic number of outputs not supported')
-    split_size = (self.type().sizes()[dim] + chunks - 1) // chunks
-    size = self.type().sizes()[dim]
+    size = sym_help._get_tensor_dim_size(self, dim)
```
Comment (neginraoof) [inline torch/onnx/symbolic_opset9.py:540]: Do we have a unit test for chunk?

File: torch/onnx/symbolic_opset9.py [hunk 8]
```diff
@@ -681,7 +690,9 @@ def leaky_relu(g, input, negative_slope, inplace=False):
 
 @parse_args('v', 'i')
 def glu(g, input, dim):
-    assert input.type().sizes()[dim] % 2 == 0
+    dim_size = sym_help._get_tensor_dim_size(input, dim)
```
Comment (neginraoof) [inline torch/onnx/symbolic_opset9.py:693]: Same question for glu: do we have a unit test?

File: torch/onnx/symbolic_opset9.py [hunk 9]
```diff
@@ -1159,11 +1173,19 @@ def log_softmax(g, input, dim, dtype=None):
 @parse_args('v', 'v', 'v', 'is', 'is', 'is', 'i', 'is', 'i', 'i', 'i', 'i', 'i')
 def _convolution(g, input, weight, bias, stride, padding, dilation,
                  transposed, output_padding, groups, benchmark, deterministic, cudnn_enabled, allow_tf32):
-    weight_size = weight.type().sizes()
+    weight_size = sym_help._get_tensor_sizes(weight)
+    try:
```
Comment (neginraoof) [inline torch/onnx/symbolic_opset9.py:1177]: Why try/catch here? Can we check if weight_size is not None and rank >= 2?

**Review by neginraoof — COMMENTED** [review]
Thanks for all the updates! Looks good. Just left a couple minor comments.

File: torch/onnx/symbolic_opset9.py (hunk 9 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:1177]: try-except is less verbose, more pythonic in this simple case.

File: torch/onnx/symbolic_opset9.py (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:540]: no, there is no explicit way of creating unsafe_unchunk. this is internally used by LSTM/GRU/RNN.

File: torch/onnx/symbolic_opset9.py (hunk 8 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:693]: added for glu

**Comment by BowenBao** [comment]
lint error is unrelated and happening for all PRs at the moment.

**Review by neginraoof — APPROVED** [review]
Looks good. Please maybe rebase for CI failures.

**Review by bzinodev — APPROVED** [review]

**Pull request closed without merging**